### ISSUE TITLE
On-device Plan

### ISSUE THREAD

**Issue by obadx** [obadx/quran-muaalem#29]
## 1. Choosing C++ over Rust  
I’m currently undecided whether to use Rust instead of C++. Most of the on‑device libraries are written in C++, and I haven’t yet found a Rust equivalent of [uniffi‑rs](https://github.com/mozilla/uniffi-rs) — a tool that lets you write core logic in Rust and export SDKs for multiple languages. The closest option I’ve found is [Scapix](https://github.com/scapix-com/scapix), but it is very limited, not actively updated, and not widely used.

## 2. Core C++ Libraries  
The on‑device ecosystem is rich with C++ projects and libraries. The most important ones for this project are:

* **[CTranslate2](https://github.com/OpenNMT/CTranslate2)** – an optimized inference engine for Transformer models, written in C++.
* **[Moonshine](https://github.com/moonshine-ai/moonshine)** – a toolkit similar to what we’re building, but more general and more complex.
* **[ONNX Runtime](https://github.com/microsoft/onnxruntime)** – a cross‑platform accelerator that can run any ONNX‑format model practically anywhere.

## 3. C++ Setup  
* **Package manager:** Investigate C++ packaging tools such as [CMake](https://cmake.org/) and [SCons](https://www.scons.org/).  
* **Compiler:** Choose between GNU g++ or LLVM Clang, and decide on a C++ standard (e.g., C++20, C++23/26).  
* **Linter:** Select a C++ linter (e.g., clang‑tidy).  
* **Formatter:** Choose a code formatter (e.g., clang‑format).  
* **Language bindings:** Define a strategy for exposing the C++ core to other languages: Kotlin (Android), Swift (iOS), WebAssembly, and Python.

## 4. Investigating the Best Model for Mobile  

We have two architectural approaches:

* **Architecture A:** A long context window (~4 seconds) with CTC heads only, trained in a streaming fashion. This can be implemented with either sliding‑window attention or full attention.  
* **Architecture B:** The approach described in [this issue](https://github.com/obadx/quran-muaalem/issues/10) — a 500 ms context with an LSTM memory for history.

We will test both approaches on a 2020‑era mobile CPU: the [Redmi 9](https://www.gsmarena.com/xiaomi_redmi_9-10233.php).

| Specification | Details |
|---------------|---------|
| **OS** | Android 10, upgradable to Android 12, MIUI 13 |
| **Chipset** | MediaTek MT6769V/CU Helio G80 (12 nm) |
| **CPU** | Octa‑core (2×2.0 GHz Cortex‑A75 & 6×1.8 GHz Cortex‑A55) |
| **GPU** | Mali‑G52 MC2 |

We will experiment with different model configurations (heads, hidden dimensions, etc.) as illustrated here: [conformer-rs usage](https://github.com/obadx/conformer-rs#usage).

```python
import torch
from conformer import ConformerConvModule

layer = ConformerConvModule(
    dim = 512,
    causal = False,             # auto-regressive or not – 1d conv will be made causal with padding if so
    expansion_factor = 2,       # what multiple of the dimension to expand for the depthwise convolution
    kernel_size = 31,           # kernel size, 17 – 31 was said to be optimal
    dropout = 0.                # dropout at the very end
)

x = torch.randn(1, 1024, 512)
x = layer(x) + x
```

### Evaluation Metrics  

For each configuration we will calculate the **Real‑Time Factor (RTF)**:

$$
\text{RTF} = \frac{\text{time taken by the model to process the context}}{\text{real‑time duration}}
$$

- For Architecture A: context length = 4 seconds  
- For Architecture B: context length = 500 ms  
- Real‑time duration = 500 ms (the stride)  

We will also log **total cores used, CPU utilization, and memory consumption**.

## 5. Training the Best Model  

1. **Prune a wav2vec2‑base (or wav2vec2‑BERT) model** down to our target size and fine‑tune it with the chosen architecture.  
2. **Before fine‑tuning:** Investigate streaming techniques with CTC.  
3. **Training phase:** Train the best architecture in two sizes — base and large — following the procedure outlined in [this issue](https://github.com/obadx/quran-muaalem/issues/10).  

## 6. Convert to C++  

* Port the [quran‑transcript](https://github.com/obadx/quran-transcript) code from Python to C++.  
* Implement the inference logic in C++.

## 7. Create an Android App for Muaalem  

* Build an **experimental** Android app that demonstrates the streaming approach in action.

**Comment by abdelkareemkobo** [comment]
Did you read LiteRT and Qualcomm ai (hub&models)?
I tried some experiments with them last week and tested the performance on my android mobile : 
The exportation  of the model and converting it into multiple language and platforms, is supported. 
- JS, Python, C++,Java..etc

### Benchamrk results

## float32, full precision, 2.4GB.
#### LiteRT CPU oneplus pad 3

Metric 	Time

Init (model load) 	4,292 ms (4.3s)

First inference 	8,455 ms (8.5s)

Average inference 	3,411 ms (3.4s)


So LiteRT on CPU gives 3.4s average vs PyTorch 81s — that's roughly 24x faster!

#### LiteRT NPU oneplus pad 3 fullback cpu

Metric 	CPU 	NNAPI

Init 	4,292 ms 	1,920 ms ✅

First inference 	8,455 ms 	5,325 ms ✅

Average inference 	3,411 ms 	3,372 ms

NNAPI gave almost no improvement on average inference — only 1% faster. This is likely because the model has many ops that fall back to CPU anyway due to unsupported op types in the NPU.

I compare the performance between core i7 gen9 intel and one plus pad 3 cpu. 

It's replacement for Tensor Lite and support PyTorch with small trick. 
It supports GPU, NPU and cpu as fallback. 

my next step was to try it in float8 or quantized version but this was requiring more ram than i have in my machine.

**Comment by obadx** [comment]
@abdelkareemkobo 

ما شاء الله يا عبد الكريم
ممكن تجرب نفس التجربة لكن ب onnx ?
وترضها علينا قبل الإجتماع
